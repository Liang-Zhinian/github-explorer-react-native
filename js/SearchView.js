/**
 * MIT License
 *
 * Copyright (c) 2017 johnwakley
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

// @flow

import React, {Component} from 'react'
import {View, FlatList} from 'react-native'
import {graphql, QueryRenderer, createPaginationContainer} from 'react-relay'
import hoistStatics from 'hoist-non-react-statics'
import {type SearchViewViewer} from './__generated__/SearchView_viewer.graphql'
import {Repo} from './exports'
import relayEnv from './common/createRelayEnvironment'
import SearchCellView from './components/SearchCellView'
import InfoBar from './components/InfoBar'

const PAGE_SIZE = 20
const FETCH_ERROR_MESSAGE = 'Failed to load repos. Pull to refresh.'

type Props = {
  viewer: SearchViewViewer,
  relay: any,
  navigation: any
}

type State = {
  isFetching: boolean,
  error: string
}

export class SearchView extends Component<void, Props, State> {
  state = {
    isFetching: false,
    error: ''
  }

  onRefresh = () => {
    if (this.props.relay.isLoading()) {
      return
    }

    this.setState({
      isFetching: true
    })

    this.props.relay.refetchConnection(PAGE_SIZE, err => {
      this.setState({
        isFetching: false
      })

      if (err) {
        console.log('Refresh ERROR: ', err)
        this.setState({error: FETCH_ERROR_MESSAGE})
      }
    })
  }

  onEndReached = () => {
    if (!this.props.relay.hasMore() || this.props.relay.isLoading()) {
      return
    }

    this.props.relay.loadMore(PAGE_SIZE, err => {
      if (err) {
        console.log('loadMore ERROR: ', err)
        this.setState({error: FETCH_ERROR_MESSAGE})
      }
    })
  }

  renderItem = ({item}: {item: any}) => {
    const {node} = item
    const primaryLanguage = node.primaryLanguage

    return (
      <SearchCellView
        title={node.name}
        description={node.description}
        language={primaryLanguage && primaryLanguage.name}
        languageColor={primaryLanguage && primaryLanguage.color}
        starCount={node.stargazers.totalCount}
        onPress={() => this.showRepoDetails(node)}
      />
    )
  }

  showRepoDetails = (repo: Repo) => {
    const {navigate} = this.props.navigation
    navigate('RepoDetail', {repo})
  }

  render () {
    const {search} = this.props.viewer

    return (
      <View style={{flex: 1, backgroundColor: 'white'}}>
        {Boolean(this.state.error) &&
          <InfoBar error autoHide>{this.state.error}</InfoBar>}
        <FlatList
          testID='repoList'
          accessibilityLabel='repoList'
          data={search.edges}
          renderItem={this.renderItem}
          keyExtractor={item => item.node.id}
          onEndReached={this.onEndReached}
          onRefresh={this.onRefresh}
          refreshing={this.state.isFetching}
          ItemSeparatorComponent={() => (
            <View style={{height: 1, backgroundColor: '#cccccc'}} />
          )}
        />
      </View>
    )
  }
}

const SearchViewPaginationContainer = createPaginationContainer(
  SearchView,
  {
    viewer: graphql`
      fragment SearchView_viewer on Query {
        search (
          first: $count, 
          after: $cursor, 
          query: $query, 
          type: REPOSITORY
        ) @connection(key: "SearchView_search") {
          pageInfo {
            hasNextPage
            endCursor
          }
          edges {
            node {
            ... on Repository {
                id,
                name,
                description,
                url,
                homepageUrl,
                license,
                owner {
                  avatarUrl,
                  login
                }
                stargazers {
                  totalCount
                },
                watchers {
                  totalCount
                },
                primaryLanguage {
                  color,
                  name
                }
              }
            }
          }
        }
      } 
    `
  },
  {
    direction: 'forward',
    getConnectionFromProps (props) {
      return props.viewer && props.viewer.search
    },
    getFragmentVariables (prevVars, totalCount) {
      return {
        ...prevVars,
        count: totalCount
      }
    },
    getVariables (props, {count, cursor}, fragmentVariables) {
      return {
        count,
        cursor,
        query: props.searchQuery
      }
    },
    variables: {cursor: null},
    query: graphql`
        query SearchViewPaginationQuery(
          $count: Int!,
          $cursor: String,
          $query: String!
        ) {
          viewer: relay {
            ...SearchView_viewer
          }
        }
      `
  }
)

const SearchViewQueryRenderer = parentProps => {
  return (
    <QueryRenderer
      environment={relayEnv}
      query={graphql`
          query SearchViewQuery(
            $count: Int!,
            $cursor: String,
            $query: String!
          ) {
            viewer: relay {
              ...SearchView_viewer
            }
          }
        `}
      variables={{
        cursor: null,
        count: PAGE_SIZE,
        query: parentProps.searchQuery
      }}
      render={({error, props}) => {
        if (props) {
          return (
            <SearchViewPaginationContainer
              viewer={props.viewer}
              navigation={parentProps.navigation}
              searchQuery={parentProps.searchQuery}
            />
          )
        } else if (error) {
          return <InfoBar error>{FETCH_ERROR_MESSAGE}</InfoBar>
        } else {
          return <InfoBar withActivityIndicator>Fetching repos</InfoBar>
        }
      }}
    />
  )
}

export default hoistStatics(SearchViewQueryRenderer, SearchView)
