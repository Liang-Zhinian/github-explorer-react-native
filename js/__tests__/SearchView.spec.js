import 'react-native'
import React from 'react'
import {SearchView} from '../SearchView'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

/*eslint-disable */
const trendingRepos = JSON.parse(
  `{
	"pageInfo": {
		"hasNextPage": true,
		"endCursor": "Y3Vyc29yOjEw"
	},
	"edges": [{
		"node": {
			"id": "MDEwOlJlcG9zaXRvcnk5MDc4NjI0Ng==",
			"name": "TypeScript-React-Starter",
			"description": "A starter template for TypeScript and React with a detailed README describing how to use the two together.",
			"url": "https://github.com/Microsoft/TypeScript-React-Starter",
			"stargazers": {
				"totalCount": 1318
			}
		}
	}, {
		"node": {
			"id": "MDEwOlJlcG9zaXRvcnk5MDk1Mzg4Nw==",
			"name": "SpreadsheetView",
			"description": "Full configurable spreadsheet view user interfaces for iOS applications. With this framework, you can easily create complex layouts like schedule, gantt chart or timetable as if you are using Excel.",
			"url": "https://github.com/kishikawakatsumi/SpreadsheetView",
			"stargazers": {
				"totalCount": 1099
			}
		}
	}]
}`
)
/*eslint-enable */

const fixture = {search: trendingRepos}

it('SearchView renders correctly', () => {
  const tree = renderer.create(<SearchView viewer={fixture} />).toJSON()
  expect(tree).toMatchSnapshot()
})
