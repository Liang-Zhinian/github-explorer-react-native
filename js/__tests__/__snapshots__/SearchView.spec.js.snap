// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchView renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    accessibilityLabel="repoList"
    alwaysBounceHorizontal={false}
    alwaysBounceVertical={true}
    data={
      Array [
        Object {
          "node": Object {
            "description": "A starter template for TypeScript and React with a detailed README describing how to use the two together.",
            "id": "MDEwOlJlcG9zaXRvcnk5MDc4NjI0Ng==",
            "name": "TypeScript-React-Starter",
            "stargazers": Object {
              "totalCount": 1318,
            },
            "url": "https://github.com/Microsoft/TypeScript-React-Starter",
          },
        },
        Object {
          "node": Object {
            "description": "Full configurable spreadsheet view user interfaces for iOS applications. With this framework, you can easily create complex layouts like schedule, gantt chart or timetable as if you are using Excel.",
            "id": "MDEwOlJlcG9zaXRvcnk5MDk1Mzg4Nw==",
            "name": "SpreadsheetView",
            "stargazers": Object {
              "totalCount": 1099,
            },
            "url": "https://github.com/kishikawakatsumi/SpreadsheetView",
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={null}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onResponderGrant={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={undefined}
    onResponderTerminationRequest={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onScrollShouldSetResponder={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    onViewableItemsChanged={undefined}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={true}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    sendMomentumEvents={false}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "overflow": "scroll",
        },
        undefined,
      ]
    }
    testID="repoList"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <RCTRefreshControl />
    <RCTScrollContentView
      collapsable={false}
      onLayout={[Function]}
      removeClippedSubviews={true}
      style={
        Array [
          false,
          undefined,
        ]
      }
    >
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel="repoCell"
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              undefined,
            ]
          }
          testID="repoCell"
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Object {
                "margin": 20,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "fontSize": 20,
                  "lineHeight": 25,
                  "marginBottom": 5,
                }
              }
            >
              TypeScript-React-Starter
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "marginBottom": 5,
                }
              }
            >
              A starter template for TypeScript and React with a detailed README describing how to use the two together.
            </Text>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "marginTop": 10,
                }
              }
            >
              <View />
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                    "paddingRight": 10,
                  }
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={false}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 16,
                      },
                      Object {
                        "color": "grey",
                        "paddingRight": 2,
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  î ¸
                </Text>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "color": "dimgrey",
                      "fontSize": 12,
                    }
                  }
                >
                  1318
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#cccccc",
            "height": 1,
          }
        }
      />
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel="repoCell"
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              undefined,
            ]
          }
          testID="repoCell"
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Object {
                "margin": 20,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "fontSize": 20,
                  "lineHeight": 25,
                  "marginBottom": 5,
                }
              }
            >
              SpreadsheetView
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "marginBottom": 5,
                }
              }
            >
              Full configurable spreadsheet view user interfaces for iOS applications. With this framework, you can easily create complex layouts like schedule, gantt chart or timetable as if you are using Excel.
            </Text>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "marginTop": 10,
                }
              }
            >
              <View />
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                    "paddingRight": 10,
                  }
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={false}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 16,
                      },
                      Object {
                        "color": "grey",
                        "paddingRight": 2,
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  î ¸
                </Text>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "color": "dimgrey",
                      "fontSize": 12,
                    }
                  }
                >
                  1099
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollContentView>
  </RCTScrollView>
</View>
`;
